Auth

YuriyHrechka opened the pull request.

MaksStud committed:
-

YuriyHrechka committed:
ADDED: users app, RegistrationLoginSerializer, RegisterLogin view, urls

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,35 @@
+from rest_framework import serializers
+
+
+class RegistrationLoginSerializer(serializers.ModelSerializer):
```
No docstring

MaksStud replied:
No docstings

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,35 @@
+    class Meta:
+        model = User
+        fields = ('username', 'password')
+        extra_kwargs = {
```
It does not need to be prescribed by you when you write a validation

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,35 @@
+        username = attrs.get('username')
+        password = attrs.get('password')
+
+        if not username or not password:
```
This will be automatically checked because the fields are required but you have turned off auto-validation

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,35 @@
+                attrs['user'] = user
+                return attrs
+            else:
+                raise serializers.ValidationError("Incorrect password")
```
You cannot return just because the password is incorrect, for security reasons

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,35 @@
+            return attrs
+
+    def create(self, validated_data):
+        user = User.objects.create_user(
```
Here, even if the user exists, it will try to create it again, it is better to use get_or _create and set the password with a separate method

MaksStud commented on bimber/users/urls.py:
```diff
@@ -0,0 +1,7 @@
+from users import views
+
+urlpatterns = [
+    path('login/', views.ApiRegisterLoginView.as_view()),
```
there must be a name= variable

MaksStud replied:
path('login/', views.ApiRegisterLoginView.as_view(), name='user-login'),

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+            user = validated_data['user']
+            token = self.get_tokens_for_user(user)
+
+            return Response({
```
You do not need to damage the message about the success of the message, the code 200 is displayed

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+
+            return Response({
+                "message": "Logged in successfully",
+                "username": user.username,
```
also don't try to return username

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+            }, status=status.HTTP_201_CREATED)
+
+    def get_tokens_for_user(self, user):
+        access_token = AccessToken.for_user(user)
```
Both Refresh and Access token should be returned

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+from rest_framework_simplejwt.tokens import AccessToken
+
+
+class ApiRegisterLoginView(APIView):
```
In general, there is a lot of unusable code

YuriyHrechka committed:
ADDED: doc strings CHANGED: serializer create and validate methods, RegisterLogin view post method abbreviated, get_tokens_for_user now returns refresh and access tokens

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,47 @@
+
+
+class RegistrationLoginSerializer(serializers.ModelSerializer):
+    """
```
Did you generate them in the gpt chat?

Here's an example

        """
        Return is physical preferences.

        :param obj: Rendezvous user object.

        :return: True if exist else False.
        """

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,47 @@
+        model = User
+        fields = ('username', 'password')
+
+    def validate(self, attrs):
```
Here you can reduce the amount of code without using get_or_create only once

MaksStud reviewed (changes requested):
You've fixed almost nothing, live carefully with the comments that are left

YuriyHrechka committed:
CHANGED: doc strings, serializer create and validate methods, RegisterLogin view name changed into ApiAuthorizationView, deleted get_tokens_for_user

YuriyHrechka committed:
ADDED: created flag

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,55 @@
+            if not user.check_password(password):
+                raise serializers.ValidationError("Username or password is incorrect")
+        attrs['user'] = user
+        attrs['created'] = created
```
Why do you need to fart a flag if you don't need to do anything else with them?

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,55 @@
+        password = attrs['password']
+
+        user, created = User.objects.get_or_create(username=username)
+        if created:
```
Too many checks, you can write
```
if created is False and user.check_password(password):
      raise serializers.ValidationError("Username or password is incorrect")
      
user.set_password(password)
user.save()
```

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,55 @@
+        :param validated_data: Dictionary with validated data containing 'user'.
+        :return: User instance.
+        """
+        return validated_data["user"]
```
It is more beautiful .get
`validated_data.get('user')`

MaksStud commented on bimber/users/urls.py:
```diff
@@ -0,0 +1,7 @@
+
+urlpatterns = [
+    path('auth/', views.ApiAuthorizationView.as_view(), name='user-auth'),
+]
```
No line at the end of the file

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+    Processes POST requests with 'username' and 'password'.
+    Returns JWT access and refresh tokens for the authenticated user.
+
+    :param request: HTTP request with 'username' and 'password'.
```
It's not a method to describe it that way, you can just leave it: "API view for user authorization."

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+        return Response({
+            'access:': str(access_token),
+            'refresh:': str(refresh)
+        }, status=status.HTTP_201_CREATED if created else status.HTTP_200_OK)
```
For safety, it is always worth returning 200 no checks required

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+    """
+    serializer_class = AuthorizationSerializer
+
+    def create(self, request, *args, **kwargs):
```
It's better to make the AuthorizationSerializer return a dictionary with tokens, so you don't have to manually override create.

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,38 @@
+from rest_framework_simplejwt.tokens import RefreshToken
+
+
+class ApiAuthorizationView(generics.CreateAPIView):
```
This name is incorrect, change it to **AuthorizationAPIView**.

MaksStud reviewed (changes requested):
The best thing to do is to have **AuthorizationSerializer** override the token dictionary and not override any methods in APIView at all.

YuriyHrechka committed:
CHANGED: AuthorizationSerializer create method now return tokens, refactored AuthorizationView

YuriyHrechka committed:
CHANGED: name of the view and added line in end of file the file

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,64 @@
+        :return: Modified attrs with 'user' key containing the user instance and 'created' key.
+        :raise: serializers.ValidationError if credentials are incorrect.
+        """
+        username = attrs['username']
```
It's better to get();

`attrs.get('username')`

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,64 @@
+        :raise: serializers.ValidationError if credentials are incorrect.
+        """
+        username = attrs['username']
+        password = attrs['password']
```
It's better to get();

attrs.get('password')

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,29 @@
+
+        :param request: HTTP request.
+        :param args: Additional positional arguments.
+        :param kwargs: Additional keyword arguments.
```
There must be an empty string

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,64 @@
+        user.refresh = refresh
+        user.access = access
+
+        return user
```
Best practics is return dict with tokes

```
return {
    "refresh": refresh
    "access": access
}
```

PS. How to not exist error use serializers.Serializer not serializers.ModelSerializer

MaksStud replied:
Not has empty line in end of file

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,29 @@
+    """
+    serializer_class = AuthorizationSerializer
+
+    def create(self, request, *args, **kwargs):
```
If serializer return dict with tokens then not rewrited create method

MaksStud commented on bimber/users/serializers.py:
```diff
@@ -0,0 +1,64 @@
+        Return the user instance from the validated data.
+
+        :param validated_data: Dictionary with validated data containing 'user'.
+        :return: User instance.
```
There must be an empty string

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,29 @@
+        return Response({
+            'access:': str(user.refresh),
+            'refresh:': str(user.access)
+        }, status=status.HTTP_200_OK)
```
Not has empty line in end of file

MaksStud commented on bimber/users/urls.py:
```diff
@@ -0,0 +1,7 @@
+
+urlpatterns = [
+    path('auth/', views.AuthorizationView.as_view(), name='user-auth'),
+]
```
Not has empty line in end of file

YuriyHrechka committed:
CHANGED: deleted access|refresh fields from serializers.py, changed doc strings, replaced [] to get, changed validation logic and create method now return JWT tokens. views.py

YuriyHrechka commented:
In the previous logic, the password was reset to the same one every time the user existed, so I changed the logic for creating and getting a user to validate method. If you don't agree, I can change it back.
Also, I changed in views inheritance of the AuthorizationView class from CreateAPIView to GenericAPIView because if I leave it empty it returns 201 code always, but you said we need return code 200.

YuriyHrechka committed:
added drf to requirements.txt, refactor validate

YuriyHrechka committed:
refactor back validate method

MaksStud commented on bimber/bimber/urls.py:
```diff
@@ -19,5 +19,6 @@
 urlpatterns = [
     path('admin/', admin.site.urls),
-    path('', include('main.urls'))
+    path('user/', include('users.urls')),
```
The file must be named bimber_user and contain the BimberUser model

MaksStud commented on bimber/users/views.py:
```diff
@@ -0,0 +1,26 @@
+from rest_framework_simplejwt.tokens import RefreshToken
+
+
+class AuthorizationView(generics.GenericAPIView):
```
No need to redefine the link and write instead
`class AuthorizationView(generics.CreateAPIView):`
